business/workloads.go

```diff
@@ -0,0 +1,531 @@
+	return pods, nil
+}
+
+func fetchWorkloads(k8s kubernetes.IstioClientInterface, namespace string, labelSelector string) (models.Workloads, error) {
+	var pods []v1.Pod
+	var repcon []v1.ReplicationController
+	var dep []v1beta1.Deployment
+	var repset []v1beta2.ReplicaSet
+	var depcon []osappsv1.DeploymentConfig
+	var fulset []v1beta2.StatefulSet
+
+	ws := models.Workloads{}
+
+	wg := sync.WaitGroup{}
+	wg.Add(6)
+	errChan := make(chan error, 6)
+
+	go func() {
+		defer wg.Done()
+		var err error
+		pods, err = k8s.GetPods(namespace, labelSelector)
+		if err != nil {
+			log.Errorf("Error fetching Pods per namespace %s: %s", namespace, err)
+			errChan <- err
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		var err error
+		dep, err = k8s.GetDeployments(namespace)
+		if err != nil {
+			log.Errorf("Error fetching Deployments per namespace %s: %s", namespace, err)
+			errChan <- err
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		var err error
+		repset, err = k8s.GetReplicaSets(namespace)
+		if err != nil {
+			log.Errorf("Error fetching ReplicaSets per namespace %s: %s", namespace, err)
+			errChan <- err
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		var err error
+		repcon, err = k8s.GetReplicationControllers(namespace)
+		if err != nil {
+			log.Errorf("Error fetching GetReplicationControllers per namespace %s: %s", namespace, err)
+			errChan <- err
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		var err error
+		if k8s.IsOpenShift() {
+			depcon, err = k8s.GetDeploymentConfigs(namespace)
+			if err != nil {
+				log.Errorf("Error fetching DeploymentConfigs per namespace %s: %s", namespace, err)
+				errChan <- err
+			}
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		var err error
+		fulset, err = k8s.GetStatefulSets(namespace)
+		if err != nil {
+			log.Errorf("Error fetching StatefulSets per namespace %s: %s", namespace, err)
+			errChan <- err
+		}
+	}()
+
+	wg.Wait()
+	if len(errChan) != 0 {
+		err := <-errChan
+		return ws, err
+	}
+
+	// Key: name of controller; Value: type of controller
+	controllers := map[string]string{}
+
+	// Find controllers from pods
+	for _, pod := range pods {
+		if len(pod.OwnerReferences) != 0 {
+			for _, ref := range pod.OwnerReferences {
+				if *ref.Controller {
+					if _, exist := controllers[ref.Name]; !exist {
+						controllers[ref.Name] = ref.Kind
+					}
+				}
+			}
+		} else {
+			if _, exist := controllers[pod.Name]; !exist {
+				// Pod without controller
+				controllers[pod.Name] = "Pod"
+			}
+		}
+	}
+
+	// Resolve ReplicaSets from Deployments
+	// Resolve ReplicationControllers from DeploymentConfigs
+	for cname, ctype := range controllers {
+		if ctype == "ReplicaSet" {
+			found := false
+			iFound := -1
+			for i, rs := range repset {
+				if rs.Name == cname {
+					iFound = i
+					found = true
+					break
+				}
+			}
+			if found && len(repset[iFound].OwnerReferences) > 0 {
+				// Delete the child ReplicaSet and add the parent controller
+				delete(controllers, cname)
```

Comment:
I'm not sure exactly what we can expect to find in the `OwnerReferences` array, but what if it's non-empty but doesn't have any `*ref.Controller` true? Maybe we should `delete(controllers, cname)` only when it's true (so, moving that line 2 lines below)

Reply:
+1
True, I used the *ref.Controller in other area and yes, it should be consistency, if not you are right, we will have a corner case when we face something odd.
I will change it.